### PR TITLE
8242152: SA does not include StackMapTables when dumping .class files

### PR DESCRIPTION
**Pull request by quadhier**
This patch adds `StackMapTable` for the class files generated by `clhsdb`'s `buildreplayjars` command. This bug manifests itself during my diagnosing [JDK-8309751](https://bugs.openjdk.org/browse/JDK-8309751) and needs to be fixed first.

I have run jtreg test `tier1-3` of release build on x86 linux finding only one failure in `tier2` caused by [JDK-8309214](https://bugs.openjdk.org/browse/JDK-8309214). `jtreg:test/hotspot/jtreg/serviceability` and `jtreg:test/jdk/sun/tools/` also passed.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8242152](https://bugs.openjdk.org/browse/JDK-8242152): SA does not include StackMapTables when dumping .class files (**Bug** - P3)

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [f2d2340d](https://git.openjdk.org/jdk/pull/14556/files/f2d2340ddc6b737454e93974570d763db4b4f9dc)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14556/head:pull/14556` \
`$ git checkout pull/14556`

Update a local copy of the PR: \
`$ git checkout pull/14556` \
`$ git pull https://git.openjdk.org/jdk.git pull/14556/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14556`

View PR using the GUI difftool: \
`$ git pr show -t 14556`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14556.diff">https://git.openjdk.org/jdk/pull/14556.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14556#issuecomment-1598614916)

### CODE REVIEW

**Comment by ashu-mehra** [comment]
@quadhier you beat me to this! Changes look good.

On another note since you mentioned [JDK-8309751](https://bugs.openjdk.org/browse/JDK-8309751) I should let you know I have a fix for it  and am planning to open a PR soon. My bad that I didn't assign that issue to myself before starting working on it.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/ConstMethod.java [hunk 1]
```diff
@@ -108,6 +109,8 @@ public ConstMethod(Address addr) {
 
   // Fields
   private static MetadataField constants;
+  // Raw stackmap data for the method (#entries + entries)
+  private static AddressField stackMapData;
```
Comment (plummercj) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/ConstMethod.java:113]: I think putting the comment on the same line as the declaration would be better

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/jcore/ClassWriter.java [hunk 2]
```diff
@@ -168,6 +169,11 @@ else if(cpConstType == JVM_CONSTANT_Long ||
         // Short deprecatedIndex = (Short) utf8ToIndex.get("Deprecated");
 
         // Code attributes
+        Short stackMapTableIndex = utf8ToIndex.get("StackMapTable");
+        _stackMapTableIndex = (stackMapTableIndex != null)?
```
Comment (plummercj) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/jcore/ClassWriter.java:173]: Add a space before the `?`

**Review by plummercj — APPROVED** [review]
I made a couple of minor suggestions. Otherwise it looks good.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/jcore/ClassWriter.java [hunk 3]
```diff
@@ -601,6 +628,17 @@ protected void writeMethod(Method m) throws IOException {
...
 
+            // write StackMapTable, if available
+            if (hasStackMapTable) {
+              writeIndex(_stackMapTableIndex);
+              dos.writeInt(stackMapAttrLen);
+              // We write bytes directly as stackMapData is
+              // raw data (#entries + entries)
+              for (int i = 0; i < stackMapData.length(); i++) {
+                dos.writeByte(stackMapData.at(i));
+              }
```
Comment (sspitsyn) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/jcore/ClassWriter.java:639]: The indent in this file is 8. So, it has to be 8 in fragments: 528-543 and 632-640.

**Review by sspitsyn — COMMENTED** [review]
This looks good in general.
I've posted one nit about formatting.
What test coverage does this update have?

**Comment by sspitsyn** [comment]
> Maybe we could add a validation in `ClhsdbDumpclass.java` to ensure that the generated class file contains some `StackMapTable`s? (As the class under validation `LingeredApp` should has `StackMapTable`s for some methods.)

It would be nice to add if it is not hard.
Thank you for the answers.

**Review by sspitsyn — APPROVED** [review]

**Review by plummercj — APPROVED** [review]

**Comment by dholmes-ora** [comment]
We see a tier1 failure on Linux-x64-debug running `serviceability/sa/ClhsdbDumpclass.java`
```
ava.lang.RuntimeException: Test ERROR java.lang.RuntimeException: 'StackMapTable:' missing from stdout/stderr
	at ClhsdbDumpclass.main(ClhsdbDumpclass.java:97)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:333)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: java.lang.RuntimeException: 'StackMapTable:' missing from stdout/stderr
	at jdk.test.lib.process.OutputAnalyzer.shouldContain(OutputAnalyzer.java:221)
	at ClhsdbDumpclass.main(ClhsdbDumpclass.java:92)
	... 4 more
```

**Comment by dholmes-ora** [comment]
What output is the test looking for i.e why is supposed to be writing it? If it is the `if (DEBUG) debugMessage(...)` calls they are not enabled.

**Comment by dholmes-ora** [comment]
I've filed https://bugs.openjdk.org/browse/JDK-8310618

**Comment by kevinjwalls** [comment]
I see the failure too.  javap needs "-v" to show the StackMap info, I see that on linux release and debug builds.

**Comment by dcubed-ojdk** [comment]
Folks the GHA for this PR reported a failure in serviceability/sa/ClhsdbDumpclass
on windows-x64 so that should have been a red flag that something was amiss.

The actual integration reported failures on windows-x64, linux-aarch64 and linux-x64
in Mach5 Tier1. Mach5 Tier3 reports a total of 14 failures on the same three platforms.
Mach5 Tier4 reports 3 failures so far on the same three platforms.

**Pull request closed without merging**